keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,578 @@
+   However, the ability to handle higher number of Pods and Jobs at a given
+   point justifies the added cost of the new API calls.
+2. Changes in the status not produced by the Job controller in
+   kube-controller-manager could affect the Job tracking. Cluster administrators
+   should make sure to protect the Job status endpoint via RBAC.
+
+## Design Details
+
+### API changes
+
+The Job status gets a new struct to hold the uncounted Pods before they are
+added to the counters.
+
+```golang
+type JobStatus struct {
+    Succeeded int32
+    Failed    int32
+    ...
+
+    // UncountedPodUIDs holds UIDs of Pods that have finished but haven't
+    // been accounted in the counters.
+    UncountedPodUIDs UncountedPodUIDs
+}
+
+// UncountedPodUIDs holds UIDs of Pods that have finished but haven't been
+// accounted in Job status counters.
+type UncountedPodUIDs struct {
+    // Enabled indicates whether Job tracking uses this struct.
+    Enabled   bool
+    // Succeeded holds UIDs of succeeded Pods.
+    Succeeded []string
+    // Succeeded holds UIDs of failed Pods.
+    Failed    []string
+}
+```
+
+### Algorithm
+
+The following algorithm updates the status counters without relying on finished
+Pods to be present indefinitely. The algorithm assumes that the Job controller
+could be stopped at any point and executed again from the first step without
+losing information. Generally, all the steps happen in a single Job sync
+cycle.
+
+1. The Job controller creates missing Pods with the finalizer
+   `batch.kubernetes.io/job-completion`
+2. The Job controller adds Pod UIDs to the `.status.uncountedPodUIDs.succeeded`
+   and `.status.uncountedPodsUIDs.failed` lists if the Pod has the
+   `batch.kubernetes.io/job-completion` finalizer and the Pod is on Succeeded
+   or Failed phase, respectively. The controller sends a status update.
+3. The Job controller removes the `batch.kubernetes.io/job-completion` finalizer
+   from all Pods on Succeeded or Failed phase. We can assume that all of them
+   have been added to the lists if the previous step succeeded.
+4. The Job controller increments the `.status.failed` and `.status.succeeded`
+   counters with the lengths of the `.status.uncountedPodsUIDs` lists, clearing
+   them afterwards. The controller sends a status update.
+   
+In the case where a user or another controller removes a Pod, which sets a
+deletion timestamp, the Job controller removes the finalizer. This is along with
+other Pods in step 3. Note that step 2 doesn't pay attention to deletion
+timestamps. So deleted Pods will be accounted in the lists and counters if they
+are in a Failed or Succeeded phase.
+   
+<<[UNRESOLVED Pod adoption ]>>
+If a Job with `.status.uncountedPodUIDs.enabled = true` can adopt a Pod, this
+might not have a finalizer. We can:
+1. Prevent adoption if the Pod doesn't have the finalizer. This is a breaking
+   change.
+1. Add the finalizer in the same patch request that modifies the owner
```

Comment:
Should go with this option.

Reply:
+1 - I don't see any disadvantages of this option.

Reply:
I don't think this should be much harder than 1.

Reply:
removed unresolved taking option 2.